switch/src/settings.cpp

```diff
@@ -0,0 +1,561 @@
+				(uint8_t*) host->rp_regist_key, &rp_regist_key_sz);
+		if(CHIAKI_ERR_SUCCESS != err)
+			CHIAKI_LOGE(this->log, "Failed to parse RP_REGIST_KEY %s (it must be a base64 encoded)", rp_regist_key_b64.c_str());
+		else
+			return true;
+	}
+	else
+		CHIAKI_LOGE(this->log, "Cannot SetHostRPKey from nullptr host");
+
+	return false;
+}
+
+void Settings::ParseFile()
+{
+	CHIAKI_LOGI(this->log, "Parse config file %s", this->filename);
+	std::fstream config_file;
+	config_file.open(this->filename, std::fstream::in);
+	std::string line;
+	std::string value;
+	bool rp_key_b, rp_regist_key_b, rp_key_type_b;
+	Host *current_host = nullptr;
+	if(config_file.is_open())
+	{
+		CHIAKI_LOGV(this->log, "Config file opened");
+		Settings::ConfigurationItem ci;
+		while(getline(config_file, line))
+		{
+			CHIAKI_LOGV(this->log, "Parse config line `%s`", line.c_str());
+			// for each line loop over config regex
+			ci = this->ParseLine(&line, &value);
+			switch(ci)
+			{
+				// got to next line
+				case UNKNOWN: CHIAKI_LOGV(this->log, "UNKNOWN config"); break;
+				case HOST_NAME:
+					CHIAKI_LOGV(this->log, "HOST_NAME %s", value.c_str());
+					// current host is in context
+					current_host = this->GetOrCreateHost(&value);
+					// all following case will edit the current_host config
+					break;
+				case HOST_IP:
+					CHIAKI_LOGV(this->log, "HOST_IP %s", value.c_str());
+					if(current_host != nullptr)
+						current_host->host_addr = value;
+
+					// reset bool flags
+					rp_key_b=false;
+					rp_regist_key_b=false;
+					rp_key_type_b=false;
+					break;
+				case PSN_ONLINE_ID:
+					CHIAKI_LOGV(this->log, "PSN_ONLINE_ID %s", value.c_str());
+					// current_host == nullptr
+					// means we are in global ini section
+					// update default setting
+					this->SetPSNOnlineID(current_host, value);
+					break;
+				case PSN_ACCOUNT_ID:
+					CHIAKI_LOGV(this->log, "PSN_ACCOUNT_ID %s", value.c_str());
+					this->SetPSNAccountID(current_host, value);
+					break;
+				case RP_KEY:
+					CHIAKI_LOGV(this->log, "RP_KEY %s", value.c_str());
+					if(current_host != nullptr)
+						rp_key_b = this->SetHostRPKey(current_host, value);
+					break;
+				case RP_KEY_TYPE:
+					CHIAKI_LOGV(this->log, "RP_KEY_TYPE %s", value.c_str());
+					if(current_host != nullptr)
+						// TODO Check possible rp_type values
+						rp_key_type_b = this->SetHostRPKeyType(current_host, value);
+					break;
+				case RP_REGIST_KEY:
+					CHIAKI_LOGV(this->log, "RP_REGIST_KEY %s", value.c_str());
+					if(current_host != nullptr)
+						rp_regist_key_b = this->SetHostRPRegistKey(current_host, value);
+					break;
+				case VIDEO_RESOLUTION:
+					this->SetVideoResolution(current_host, value);
+					break;
+				case VIDEO_FPS:
+					this->SetVideoFPS(current_host, value);
+					break;
+			} // ci switch
+			if(rp_key_b && rp_regist_key_b && rp_key_type_b)
+				// the current host contains rp key data
+				current_host->rp_key_data = true;
```

Comment:
hi, 
I copy pasted the config and it seems to work.
what line termination are you using? 
do you have any steps to reproduce the issue ?

I still investigating this problem

Reply:
I didn't try this config on the switch directly. It happens when I launch it on Linux with no config file existing, then select "Account ID". Then the GUI hangs when it is normally supposed to show the keyboard (not sure if this is a bug or limitation of borealis). After killing it with SIGINT and restarting, it segfaults.

Reply:
OK I see,
sorry I did not inform you about this.

from the nintendo switch there is a built-in function that show a keyboad overlay.
the linux version is just for testing and validation. 

for linux you have to use the shell's stdin to push your inputs.

there is a similar issue with sdl controller events.
all are mapped with the switch keys in mind. 
if you try to use a controller on linux all imputes are scrambled

> After killing it with SIGINT and restarting, it segfaults. 
very interesting, I 'm checking

Reply:
I'm sorry, but I failed to reproduce the issue.
I tried with SIGKILL SIGINT ... with trash in the stdin ... and with trash in chiaki.conf.
the SIGINT close the stdin and write the file locally.
https://github.com/H0neyBadger/chiaki/blob/019d7df856d0c724d221c342722b78c650f79d6b/switch/src/io.cpp#L338
there is no verification for the linux version.  
is this 100% reproducible ?

to be honest, I'm not proud of the settings class.
it's "hand written/homemade" and may lead to many issues.

I m torn between two solution about this config/settings class.
1) use an external lib to manage config file. (but it makes the chiaki project heavier)
2) create a custom class to save our rp key. (our project remains dependency free and easier to port. but on the other hand it contains more bugs) 

let me know if you have any suggestion

Reply:
Yeah it seems to be 100% reproducible for me. The crash happens only when I re-open the application the second time, not after the signal. `current_host` is null there. But this shouldn't block merging the pr. I can also try to fix it locally since I can reproduce it, just didn't check it in detail yet.

About the current settings implementation, I am not a huge fan of it either. In the Android Version, there is the possibility to import/export settings from/to a json-based format: https://github.com/thestr4ng3r/chiaki/blob/master/android/app/src/main/java/com/metallic/chiaki/common/SerializedSettings.kt
The Qt GUI doesn't have that yet, but it's planned there too. The idea is that you have an easy way of transferring the settings between all the different ports, so the switch version should eventually get that as well. In that sense, we will need a json parser sooner or later anyway, so we could also use it for the settings here, we just should be careful in which json parser library we pick since there are many of them with different properties.
Maybe it would even be possible to make the settings in the same format as the common json so you can just drop a json from your PC on the SD card and have all your registered consoles available.

Reply:
I aggre, portable and standard config is definitely a good design 

currently borealis is including a copy of the https://github.com/nlohmann/json
https://github.com/natinusala/borealis/blob/master/library/include/borealis/extern/nlohmann/json.hpp

I'm ok to follow your advice. just let me know if you have any structure or standard in mind. I will do my best to integrate quickly (but safely).